### PR TITLE
Ensuring we don't track views of author or unpublished

### PR DESCRIPTION
**Pull request by jeremyf**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [?] Feature
- [?] Bug Fix
- [x] Optimization

## Description

Prior to this commit, I was surprised to learn that we:

1) Tracked an author's view of their article.
2) Tracked views of an unpublished article.

This came up from a Forem creator asking if they could reset the view
counter.  Or trigger the reset on publication.

I think a general business logic policy of don't track views for the
author and don't track views for unpublished articles is a reasonable
default.

Were we to pursue the clear views on publication, we'd need to consider
something that went from unpublished -> published -> unpublished ->
published.  Without a more explicit state machine, triggering a
busineiss logic behavior seems a bit unexpected.

In other words, I wrote an article.  There are 20 views when I realize
that I need to unpublished it.  I make the changes in the unpublished
state, and re-publish.  I'd assume that those 20 views would still be
"recorded" and counted towards my article's view counts.

## Related Tickets & Documents

Loosely related to #16140, #16142.

## QA Instructions, Screenshots, Recordings

None.

### UI accessibility concerns?

None.

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [X] I will share this change internally with the appropriate teams

### CODE REVIEW

**Comment by jeremyf** [comment]
This is up in a draft status as I get further information from @amywtlin

**Comment by jeremyf** [comment]
From Amy:

> We shouldn't be counting & recording stats when posts are unpublished.

File: app/workers/articles/update_page_views_worker.rb [hunk 1]
```diff
@@ -2,16 +2,18 @@ module Articles
   class UpdatePageViewsWorker
     include Sidekiq::Worker
 
-    sidekiq_options queue: :medium_priority,
+    sidekiq_options(queue: :medium_priority,
```
Comment (jeremyf) [inline app/workers/articles/update_page_views_worker.rb:5]: I added parenthesis around this as my buffer parser was insisting on some odd indentation for documentation.

File: app/controllers/page_views_controller.rb [hunk 2]
```diff
@@ -31,14 +31,7 @@ def create
...
   def update
-    if session_current_user_id
-      page_view = PageView.order(created_at: :desc)
-        .find_or_create_by(article_id: params[:id], user_id: session_current_user_id)
-
-      unless page_view.new_record?
-        page_view.update_column(:time_tracked_in_seconds, page_view.time_tracked_in_seconds + 15)
-      end
-    end
+    Articles::PageViewUpdater.call(article_id: params[:id], user_id: session_current_user_id) if session_current_user_id
```
Comment (citizen428) [inline app/controllers/page_views_controller.rb:34]: Nitpick (style): Maybe it's just me but this modifier `if` is a bit far to the right of this line. I think I'd prefer a regular `if` here. Or maybe even add an early return in `Articles::PageViewUpdater` so we can call it unconditionally and the controller doesn't need to know anything about page view logic.

**Review by citizen428 — APPROVED** [review]

File: app/services/articles/page_view_updater.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  #      views.
+  module PageViewUpdater
+    # @param article_id [Integer]
+    # @param user_id [Integer]
+    #
+    # @return [TrueClass] if we updated or created a PageView
+    # @return [FalseClass] if we did not update a PageView
+    #
+    # @note Regardless of return status, consider the business logic successful unless we raise an
+    #       exception.  The return value is present for easing testing.  The `find_or_create_by`
```
Comment (Ridhwana) [inline app/services/articles/page_view_updater.rb:14]:
```suggestion
    #       exception. The return value is present for easing testing. The `find_or_create_by`
```

**Review by Ridhwana — APPROVED** [review]
Code looks good 🚀  Just suggested the removal of some extra whitespace in a comment.

File: app/services/articles/page_view_updater.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    #       adds a complication in the testing logic
+    #       (e.g., `expect { Articles::PageViewUpdater.call }.not_to change(PageView, :count) `
+    def self.call(article_id:, user_id:)
+      # Don't record views to unpublished articles.
+      return false if Article.unpublished.exists?(id: article_id)
+      # Don't record author's own views.
+      return false if Article.published.exists?(id: article_id, user_id: user_id)
```
Comment (jgaskins) [inline app/services/articles/page_view_updater.rb:21]: Nice solution to this! Avoiding instantiating the object or even pulling back the full `articles` row! 💯

**Review by jgaskins — APPROVED** [review]

File: app/services/articles/page_view_updater.rb (hunk 4 above)
Comment (jeremyf) [reply app/services/articles/page_view_updater.rb:21]: I can do it in one query if we want.  It just looked uglier by a lot.

File: app/services/articles/page_view_updater.rb (hunk 3 above)
Comment (jeremyf) [reply app/services/articles/page_view_updater.rb:14]: For sentences I always use two spaces between periods.  This is because my text editor allows for navigation by sentence.  So my strong preference is to keep two spaces.

Other use cases is that I can quickly expand my highlighted context, starting from cursor, to word, to sentence, to paragraph, to buffer.

Comment (citizen428) [reply app/services/articles/page_view_updater.rb:14]: While I have no problems with two spaces after full stops (I'm old enough to remember that this used to be standard at one point) I'm not super convinced by this argument:

> This is because my text editor allows for navigation by sentence. So my strong preference is to keep two spaces.

This is a codebase shared by many people, IMHO we can't optimize for everyone's preferred tool choice.
